programs/stake/src/stake_state.rs

```diff
@@ -6409,6 +6455,236 @@ mod tests {
+            ),
+        ] {
+            let rent = Rent::default();
+            let rent_exempt_reserve = rent.minimum_balance(std::mem::size_of::<StakeState>());
+            let stake_pubkey = Pubkey::new_unique();
+            let signers = HashSet::from([stake_pubkey]);
+            let meta = Meta {
+                rent_exempt_reserve,
+                ..Meta::auto(&stake_pubkey)
+            };
+
+            for stake_state in &[
+                StakeState::Initialized(meta),
+                StakeState::Stake(meta, Stake::default()),
+                StakeState::Stake(meta, just_stake(stake_delegation)),
+            ] {
+                let stake_account = AccountSharedData::new_ref_data_with_space(
+                    stake_delegation + rent_exempt_reserve,
+                    stake_state,
+                    std::mem::size_of::<StakeState>(),
+                    &id(),
+                )
+                .unwrap();
+                let stake_keyed_account = KeyedAccount::new(&stake_pubkey, true, &stake_account);
+
+                let vote_pubkey = Pubkey::new_unique();
+                let vote_account = RefCell::new(vote_state::create_account(
+                    &vote_pubkey,
+                    &Pubkey::new_unique(),
+                    0,
+                    100,
+                ));
+                let vote_keyed_account = KeyedAccount::new(&vote_pubkey, false, &vote_account);
+
+                assert_eq!(
+                    expected_result,
+                    stake_keyed_account.delegate(
+                        &vote_keyed_account,
+                        &Clock::default(),
+                        &StakeHistory::default(),
+                        &Config::default(),
+                        &signers,
+                    ),
+                );
+            }
+        }
+    }
+
+    /// Ensure that `withdraw()` respects the MINIMUM_STAKE_DELEGATION requirements
+    /// - Assert 1: withdrawing so remaining stake is equal-to the minimum is OK
+    /// - Assert 2: withdrawing so remaining stake is less-than the minimum is not OK
+    #[test]
+    fn test_withdraw_minimum_stake_delegation() {
+        let starting_stake_delegation = MINIMUM_STAKE_DELEGATION;
+        for (ending_stake_delegation, expected_result) in [
+            (MINIMUM_STAKE_DELEGATION, Ok(())),
+            (
+                MINIMUM_STAKE_DELEGATION - 1,
+                Err(InstructionError::InsufficientFunds),
+            ),
+        ] {
+            let rent = Rent::default();
+            let rent_exempt_reserve = rent.minimum_balance(std::mem::size_of::<StakeState>());
+            let stake_pubkey = Pubkey::new_unique();
+            let meta = Meta {
+                rent_exempt_reserve,
+                ..Meta::auto(&stake_pubkey)
+            };
+
+            for stake_state in &[
+                StakeState::Initialized(meta),
+                StakeState::Stake(meta, just_stake(starting_stake_delegation)),
```

Comment:
This case is interesting, because we could have a stake whose delegation is currently less than `MINIMUM_STAKE_DELEGATION`, but also has some non-staked lamports that bring it up above the minimum.  If they try to withdraw those extra lamports, they'll be capped too.

All that to say, this works, well done!